translator/trace/jaegerthrift_to_protospan.go

```diff
@@ -0,0 +1,240 @@
+		Link: make([]*tracepb.Span_Link, 0, len(jrefs)),
+	}
+
+	for _, jref := range jrefs {
+		var linkType tracepb.Span_Link_Type
+		if jref.RefType == jaeger.SpanRefType_CHILD_OF {
+			linkType = tracepb.Span_Link_CHILD_LINKED_SPAN
+		} else {
+			// SpanRefType_FOLLOWS_FROM doesn't map well to OC, so treat all other cases as unknown
+			linkType = tracepb.Span_Link_TYPE_UNSPECIFIED
+		}
+
+		link := &tracepb.Span_Link{
+			TraceId: jTraceIDToOCProtoTraceID(jref.TraceIdHigh, jref.TraceIdLow),
+			SpanId:  jSpanIDToOCProtoSpanID(jref.SpanId),
+			Type:    linkType,
+		}
+		links.Link = append(links.Link, link)
+	}
+
+	return links
+}
+
+func jTraceIDToOCProtoTraceID(high, low int64) []byte {
+	traceID := make([]byte, 16)
+	binary.BigEndian.PutUint64(traceID[:8], uint64(high))
+	binary.BigEndian.PutUint64(traceID[8:], uint64(low))
+	return traceID
+}
+
+func jSpanIDToOCProtoSpanID(id int64) []byte {
+	spanID := make([]byte, 8)
+	binary.BigEndian.PutUint64(spanID, uint64(id))
+	return spanID
+}
+
+func jtagsToAttributes(tags []*jaeger.Tag) (tracepb.Span_SpanKind, *tracepb.Span_Attributes) {
+	if tags == nil {
+		return tracepb.Span_SPAN_KIND_UNSPECIFIED, nil
+	}
+
+	var sKind tracepb.Span_SpanKind
+	sAttribs := &tracepb.Span_Attributes{
+		AttributeMap: make(map[string]*tracepb.AttributeValue, len(tags)),
+	}
+
+	for _, tag := range tags {
+		// Take the opportunity to get the "span.kind" per OpenTracing spec, however, keep it also on the attributes.
+		// TODO: Q: Replace any OpenTracing literals by importing github.com/opentracing/opentracing-go/ext?
+		if tag.Key == "span.kind" {
+			switch tag.GetVStr() {
+			case "client":
+				sKind = tracepb.Span_CLIENT
+			case "server":
+				sKind = tracepb.Span_SERVER
+			}
+		}
+
+		attrib := &tracepb.AttributeValue{}
+		switch tag.GetVType() {
+		case jaeger.TagType_STRING:
+			attrib.Value = &tracepb.AttributeValue_StringValue{
+				StringValue: &tracepb.TruncatableString{Value: tag.GetVStr()},
+			}
+		case jaeger.TagType_DOUBLE:
+			attrib.Value = &tracepb.AttributeValue_DoubleValue{
+				DoubleValue: tag.GetVDouble(),
+			}
+		case jaeger.TagType_BOOL:
+			attrib.Value = &tracepb.AttributeValue_BoolValue{
+				BoolValue: tag.GetVBool(),
+			}
+		case jaeger.TagType_LONG:
+			attrib.Value = &tracepb.AttributeValue_IntValue{
+				IntValue: tag.GetVLong(),
+			}
+		case jaeger.TagType_BINARY:
+			attrib.Value = &tracepb.AttributeValue_StringValue{
+				StringValue: &tracepb.TruncatableString{Value: base64.StdEncoding.EncodeToString(tag.GetVBinary())},
+			}
+		default:
+			attrib.Value = &tracepb.AttributeValue_StringValue{
+				StringValue: &tracepb.TruncatableString{Value: fmt.Sprintf("<Unknown Jaeger TagType %q>", tag.GetVType())},
+			}
+		}
+
+		sAttribs.AttributeMap[tag.Key] = attrib
+	}
+
+	return sKind, sAttribs
+}
+
+// epochMicrosecondsAsTime converts microseconds since epoch to time.Time value.
+func epochMicrosecondsAsTime(ts uint64) time.Time {
+	seconds := ts / 1000000
```

Comment:
Debugging and reading this is a lot easier if we use `1en` e.g. `seconds := us / 1e6`
so let's make the argument name `us`
```go
func epochMicrosecondsAsTime(ts uint64) time.Time {
   seconds := us / 1e6
   nanos := 1e3  * (us % 1e6)
   return
```
but also I believe the direct and cheaper solution is
```go
func epochMicrosecondsAsTime(us uint64) time.Time {
   return time.Unix(0, int64(us*1e3))
}
```
since there are 1e3 nanoseconds in a microsecond but please help me check my solution

Reply:
It didn't occur to me that it could take zero for the seconds, but makes sense that it does; nice.